modules/detection/test/testAprilTag.cpp

```diff
@@ -725,9 +725,12 @@ TEST_CASE("Apriltag getTagsPoints3D test", "[apriltag_get_tags_points3D_test]")
     pose.computePose(vpPose::DEMENTHON_VIRTUAL_VS, cMo_manual);
 
     const vpHomogeneousMatrix& cMo = cMo_vec[i];
+    double epsilon = 1e-9;
+    // Note that using epsilon = std::numeric_limits<double>::epsilon() makes this test
+    // failing when none of the Lapack 3rd party libraries, nor the built-in are used
     for (unsigned int row = 0; row < cMo.getRows(); row++) {
       for (unsigned int col = 0; col < cMo.getCols(); col++) {
-        CHECK(vpMath::equal(cMo[row][col], cMo_manual[row][col], std::numeric_limits<double>::epsilon()));
+        CHECK(vpMath::equal(cMo[row][col], cMo_manual[row][col], epsilon));
```

Comment:
Looking at it quicky, this does not seem normal.

We are comparing the cMo estimated using `DEMENTHON_VIRTUAL_VS` inside the `vpDetectorAprilTag` class and manually (getting the 2D/3D points and calling vpPose `DEMENTHON_VIRTUAL_VS` manually).

The result should be strictly identical, it should be equivalent to calling the vpPose function two times with the same input (to be checked if it is really the case?), and yielding a slightly different result.

I will try to investigate.

Reply:
Ok, I did not read the comment above.

I would prefer to keep the `std::numeric_limits<double>::epsilon()` threshold and add an exception for this specific case (but I am not able to reproduce it).

---

BTW, I am testing in a Docker container (Ubuntu 16.04) without any BLAS or LAPACK libraries, and turn the `WITH_LAPACK` variable to false.

<details><summary>ViSP-third-party.txt</summary>
<p>

```
==========================================================
General configuration information for ViSP 3.3.1

  Version control:               unknown

  Platform:
    Timestamp:                   2020-04-02T22:01:08Z
    Host:                        Linux 4.15.0-91-generic x86_64
    CMake:                       3.5.1
    CMake generator:             Unix Makefiles
    CMake build tool:            /usr/bin/make
    Configuration:               Release

  C/C++:
    Built as dynamic libs?:      yes
    C++ Compiler:                /usr/bin/c++  (ver 5.4.0)
    C++ flags (Release):         -Wall -Wextra -fopenmp -std=c++11 -fvisibility=hidden -mavx -fPIC -O3 -DNDEBUG
    C++ flags (Debug):           -Wall -Wextra -fopenmp -std=c++11 -fvisibility=hidden -mavx -fPIC -g
    C Compiler:                  /usr/bin/cc
    C flags (Release):           -Wall -Wextra -fopenmp -std=c++11 -fvisibility=hidden -mavx -fPIC -O3 -DNDEBUG
    C flags (Debug):             -Wall -Wextra -fopenmp -std=c++11 -fvisibility=hidden -mavx -fPIC -g
    Linker flags (Release):
    Linker flags (Debug):

  ViSP modules:
    To be built:                 core gui imgproc io java_bindings_generator klt me sensor ar blob robot visual_features vs vision detection mbt tt tt_mi
    Disabled:                    -
    Disabled by dependency:      -
    Unavailable:                 java

  Python (for build):            /usr/bin/python2.7

  Java:                          
    ant:                         NO
    JNI:                         NO

  Build options: 
    Build deprecated:            yes
    Build with moment combine:   no

  Mathematics: 
    Blas/Lapack:                 no
    \- Use MKL:                  no
    \- Use OpenBLAS:             no
    \- Use Atlas:                no
    \- Use Netlib:               no
    \- Use GSL:                  no
    \- Use Lapack (built-in):    no
    Use Eigen3:                  no
    Use OpenCV:                  no

  Simulator: 
    Ogre simulator: 
    \- Use Ogre3D:               no
    \- Use OIS:                  no
    Coin simulator: 
    \- Use Coin3D:               no
    \- Use SoWin:                no
    \- Use SoXt:                 no
    \- Use SoQt:                 no
    \- Use Qt5:                  no
    \- Use Qt4:                  no
    \- Use Qt3:                  no

  Media I/O: 
    Use JPEG:                    no
    Use PNG:                     no
    \- Use ZLIB:                 no
    Use OpenCV:                  no
    Use stb_image (built-in):    yes (ver 2.22.0)

  Real robots: 
    Use Afma4:                   no
    Use Afma6:                   no
    Use Franka:                  no
    Use Viper650:                no
    Use Viper850:                no
    Use aria (Pioneer):          no
    Use PTU46:                   no
    Use Biclops PTU:             no
    Use Flir PTU SDK:            no
    Use Parrot ARSDK:            no
    \-Use ffmpeg:                no
    Use Virtuose:                no
    Use qbdevice (built-in):     yes (ver 2.6.0)

  GUI: 
    Use X11:                     yes
    Use GTK:                     no
    Use OpenCV:                  no
    Use GDI:                     no
    Use Direct3D:                no

  Cameras: 
    Use DC1394-2.x:              no
    Use CMU 1394:                no
    Use V4L2:                    no
    Use directshow:              no
    Use OpenCV:                  no
    Use Flycapture:              no
    Use Pylon:                   no

  RGB-D sensors: 
    Use Realsense:               no
    Use Realsense2:              no
    Use Kinect:                  no
    \- Use libfreenect:          no
    \- Use libusb-1:             no
    \- Use pthread:              yes
    Use PCL:                     no
    \- Use VTK:                  no

  F/T sensors: 
    Use atidaq (built-in):       no
    Use comedi:                  no
    Use IIT SDK:                 no

  Detection: 
    Use zbar:                    no
    Use dmtx:                    no
    Use AprilTag (built-in):     yes (ver 3.1.1)
    \- Use AprilTag big family:  no

  Misc: 
    Use Clipper (built-in):      no
    Use pugixml (built-in):      yes (ver 1.9.0)
    Use libxml2:                 no

  Optimization: 
    Use OpenMP:                  yes
    Use pthread:                 yes
    Use pthread (built-in):      no
    Use cxx standard:            11

  Documentation: 
    Use doxygen:                 no

  Tests and samples:
    Use catch2 (built-in):       yes (ver 2.9.2)
    Tests:                       yes
    Demos:                       no
    Examples:                    no
    Tutorials:                   no

  Install path:                  /usr/local

==========================================================
```

</p>
</details>

I am getting this error:

<details><summary>./testAprilTag</summary>
<p>

```

Apriltag family: 36h11
	Pose estimation method: HOMOGRAPHY
		Same size, Tag: 0
			Translation error: 0.00702138 / Rotation error: 0.0228843
		Same size, Tag: 1
			Translation error: 0.000924979 / Rotation error: 0.00172373
		Same size, Tag: 2
			Translation error: 0.0184756 / Rotation error: 0.0213708
		Custom size, Tag: 0
			Translation error: 0.00702138 / Rotation error: 0.0228843
		Custom size, Tag: 1
			Translation error: 0.000924979 / Rotation error: 0.00172373
		Custom size, Tag: 2
			Translation error: 0.0184756 / Rotation error: 0.0213708
		Custom size, Tag: 3
			Translation error: 0.00445275 / Rotation error: 0.0168339
		Custom size, Tag: 4
			Translation error: 0.00727533 / Rotation error: 0.0182167
		Custom size + aligned Z-axis, Tag: 0
			Translation error: 0.00702138 / Rotation error: 0.0185626
		Custom size + aligned Z-axis, Tag: 1
			Translation error: 0.000924979 / Rotation error: 0.00188652
		Custom size + aligned Z-axis, Tag: 2
			Translation error: 0.0184756 / Rotation error: 0.0174533
		Custom size + aligned Z-axis, Tag: 3
			Translation error: 0.00445275 / Rotation error: 0.0217513
		Custom size + aligned Z-axis, Tag: 4
			Translation error: 0.00727533 / Rotation error: 0.0250883
	Pose estimation method: HOMOGRAPHY_VIRTUAL_VS

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
testAprilTag is a Catch v2.9.2 host application.
Run with -? for options

-------------------------------------------------------------------------------
Apriltag pose estimation test
-------------------------------------------------------------------------------
/workspace/visp/modules/detection/test/testAprilTag.cpp:130
...............................................................................

/workspace/visp/modules/detection/test/testAprilTag.cpp:130: FAILED:
  {Unknown expression after the reported line}
due to unexpected exception with message:
  Cannot compute pseudo-inverse. Install Lapack, Eigen3, OpenCV or GSL 3rd
  party


Apriltag family: 36h11
	id: 0 - RMSE: 0.00028054
	id: 1 - RMSE: 0.000349587
	id: 2 - RMSE: 0.000295341
	id: 3 - RMSE: 0.000150511
	id: 4 - RMSE: 0.000357982

Apriltag family: 25h9
	id: 0 - RMSE: 0.000205857
	id: 1 - RMSE: 0.000282345
	id: 2 - RMSE: 0.00023119
	id: 3 - RMSE: 0.000163054
	id: 4 - RMSE: 0.000308641

Apriltag family: 16h5
	id: 0 - RMSE: 0.000316034
	id: 1 - RMSE: 0.00036884
	id: 2 - RMSE: 0.000330228
	id: 3 - RMSE: 0.00027399
	id: 4 - RMSE: 0.000397781

Apriltag family: CIRCLE21h7
	id: 0 - RMSE: 0.000257834
	id: 1 - RMSE: 0.00027308
	id: 2 - RMSE: 0.000275198
	id: 3 - RMSE: 0.000192135
	id: 4 - RMSE: 0.000201747
-------------------------------------------------------------------------------
Apriltag regression test
-------------------------------------------------------------------------------
/workspace/visp/modules/detection/test/testAprilTag.cpp:434
...............................................................................

/workspace/visp/modules/detection/test/testAprilTag.cpp:434: FAILED:
  {Unknown expression after the reported line}
due to unexpected exception with message:
  Cannot compute pseudo-inverse. Install Lapack, Eigen3, OpenCV or GSL 3rd
  party

-------------------------------------------------------------------------------
Apriltag copy constructor test
-------------------------------------------------------------------------------
/workspace/visp/modules/detection/test/testAprilTag.cpp:533
...............................................................................

/workspace/visp/modules/detection/test/testAprilTag.cpp:533: FAILED:
  {Unknown expression after the reported line}
due to unexpected exception with message:
  Cannot compute pseudo-inverse. Install Lapack, Eigen3, OpenCV or GSL 3rd
  party

-------------------------------------------------------------------------------
Apriltag assignment operator test
-------------------------------------------------------------------------------
/workspace/visp/modules/detection/test/testAprilTag.cpp:599
...............................................................................

/workspace/visp/modules/detection/test/testAprilTag.cpp:599: FAILED:
  {Unknown expression after the reported line}
due to unexpected exception with message:
  Cannot compute pseudo-inverse. Install Lapack, Eigen3, OpenCV or GSL 3rd
  party

-------------------------------------------------------------------------------
Apriltag getTagsPoints3D test
-------------------------------------------------------------------------------
/workspace/visp/modules/detection/test/testAprilTag.cpp:665
...............................................................................

/workspace/visp/modules/detection/test/testAprilTag.cpp:665: FAILED:
  {Unknown expression after the reported line}
due to unexpected exception with message:
  Cannot compute pseudo-inverse. Install Lapack, Eigen3, OpenCV or GSL 3rd
  party

===============================================================================
test cases:   6 |   1 passed | 5 failed
assertions: 137 | 132 passed | 5 failed
```

</p>
</details>

The error makes much sense, and I would prefer just keeping the original threshold `std::numeric_limits<double>::epsilon()`.

Reply:
1. I checked again that I'm able to reproduce the `testAprilTag` failure when Lapack is not used (`WITH_LAPACK=OFF` and all the `find_package()` were commented in `ChooseBLAS.cmake`). The smallest epsilon that makes the test successful is 1e-14. I will add an exception for this particular case:
    ```
    // Note that using epsilon = std::numeric_limits<double>::epsilon() makes this test
    // failing on Ubuntu 18.04 when none of the Lapack 3rd party libraries, nor the built-in are used.
    // Admissible espilon value is 1e-14. Using 1e-15 makes the test failing.
    #ifdef VISP_HAVE_LAPACK
        double epsilon = std::numeric_limits<double>::epsilon();
    #else
        double epsilon = 1e-12;
    #endif
    ```

2. When none of the 3rd parties (Lapack, OpenCv and Eigen3) are used, there is not only `testAprilTag` that fails. I will do the changes in an other PR after merging this one

Reply:
- I am ok as soon as we don't relax the threshold for the common cases (ViSP is built with MKL, OpenBLAS, or with built-in Lapack).
- So, when `WITH_LAPACK=OFF` and without `USE_BLAS`, it should be either `Eigen` or `OpenCV` that perform the calculation? I still don't see why we have this issue when Eigen or OpenCV is used, but for now it is fine since there is no regression.